doc: rgw adding a very basic multisite doc

theanalyst opened the pull request:
Very basic documentation for the new multisite features, based on emails in the ceph-devel list
TODO
- Hyperlinks within and updating index
- Configuration and installation guide

cbodley commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,487 @@
+Note that every realm has an id, which allows for flexibility like renaming the
+realm later, should the need arise. The ``current_period`` changes whenever we
+change anything in the master zone. The ``epoch`` is incremented when there's a
+change in configuration which doesn't modify the master zone
```
> The ``current_period`` changes whenever we change anything in the master zone. The ``epoch`` is incremented when there's a change in configuration which doesn't modify the master zone

this is a bit off, from my understanding. any change to the configuration that doesn't change which zone is the master_zone will just increment the _period's_ epoch. on the other hand, changing the master_zone to point to a different zone will a) generate a new period id with period epoch=1, b) assign realm's current_period to the new period's id, and c) increment the realm's epoch. be careful not to confuse the realm's epoch with the period's epoch

theanalyst replied:
Thanks for clarifying, I'll update

yehudasa commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,487 @@
+
+- **Realm**: A realm is a container for zonegroups, this allows for separation
+  of zonegroups themselves between clusters. It is possible to create multiple
+  realm, making it easier to run completely different configurations in the same
```
multiple realms

oritwas commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,487 @@
+
+Configure a realm called ``gold``, which for ::
+
+  # radosgw-admin realm create --rgw-realm=gold
```
I would add the --default flag to set the realm as default so you don't need to provide in all the admin commands

theanalyst replied:
Sure makes sense, I'll change that

oritwas commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,487 @@
+
+::
+
+   # radosgw-admin zonegroup create --rgw-zonegroup=us --endpoints=http://rgw1:80 --master
```
also add --default

oritwas commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,487 @@
+    "realm_id": "4a367026-bd8f-40ee-b486-8212482ddcd7"
+    }
+
+Now we make this as the default zonegroup via the following command ::
```
you can add the flag --default when creating the zonegroup

theanalyst replied:
ack

oritwas commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,487 @@
+
+  # radosgw-admin zonegroup add --rgw-zonegroup=us --rgw-zone=us-east
+
+
```
We allow adding the zone to the zonegroup in the create command by adding --rgw-zonegroup=<zg> and setting it as default by adding --default.

oritwas commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,487 @@
+
+  # radosgw-admin zonegroup default --rgw-zonegroup=us
+
+Next we create the new zone, ``us-west``, with the same system keys::
```
Just to make it simpler, I would have created the secondary zone on the master zone and than would do just one period update --commit.

theanalyst replied:
Hm ok, I didn't know of this approach, sounds simpler, I'll update teh docs

theanalyst replied:
Not able to get this working completely yet, the secondary ceph cluster does see the realm, period and zonegroups, however a zone list still turns up empty

cbodley replied:
@oritwas I looked into making this work (creating a zone on the primary cluster and pulling it to the secondary), but I don't think it's actually possible with our design. 'zone create' will create a local `RGWZoneParams` object that contains all of the pool names, access key, etc. but the period only stores them as `RGWZone` - which is an external representation that tells other zones how to connect to/use it. this means that a 'period pull' can't get the `RGWZoneParams` from one cluster to another

so zones need to be created on their own cluster, for now. we might consider adding a 'zone pull' command, but I'm not sure it's a good idea..

oritwas replied:
@cbodley , the problem is that the zone id will be different. Maybe store the RGWZoneParams in the period or create the zone when doing period pull?

oritwas replied:
@theanalyst, after discussing it with @cbodley:
If we have two different clusters:
on the master: realm creation, zonegroup creation, master zone creation, period update --commit
on the second zone: realm pull, period pull, create second zone, period update --comitt
For one cluster we can have a simpler way of configuration as everything can be done on the master gateway including the second zone creation (no need for realm pull and period pull).

theanalyst replied:
@oritwas @cbodley 
> If we have two different clusters:
on the master: realm creation, zonegroup creation, master zone creation, period update --commit
on the second zone: realm pull, period pull, create second zone, period update --comitt
For one cluster we can have a simpler way of configuration as everything can be done on the master gateway including the second zone creation (no need for realm pull and period pull). 

Alright, so if its a single cluster the second approach works,. sounds useful.. and for 2 different clusters the std. approach ... thanks! I'll try this locally once and update the docs

oritwas commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,487 @@
+  }
+
+  # radosgw-admin realm default --rgw-realm=gold
+
```
you are missing period pull

theanalyst commented:
thanks for the reviews, pushed some commits addressing the comments, (can be squashed later if necessary). 

@oritwas In the last commit cfee62a0e79218729fc6d0f5135a9f8b5366975a I tried updating the doc to create the secondary zone from the first cluster and pull the period, realm etc. see if this makes sense?

theanalyst commented:
@oritwas @cbodley I've updated the doc with your comments, changed the doc a bit to show both the possibilities, ie. zone1 and zone2 in same cluster, zone2 in different clusters. Do take a look

jrosenboom commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,710 @@
+environment. For eg., something like::
+
+  $ SYSTEM_ACCESS_KEY=1555b35654ad1656d805
+  $ SYSTEM_SECRET_KEY=h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q==
```
Can you add an example how to create these keys? Would be bad if people just use the sample from the docs because it is too complicated to find out how to build your own.

theanalyst replied:
As far as ceph is concerned, I believe we may not actually have an alphanumeric restriction,  but for the sake of uniformity, we can probably generate keys by getting 20 random alphanumeric chars for access, and 40 for secret

`cat /dev/urandom | tr -dc 'a-zA-Z0-9 | fold -w 20  | head -n1' and 40 for secret is what comes to my mind.. my bash-fu isn't that great and any alternatives are welcome :-)

jrosenboom commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,710 @@
+
+- **Zonegroup**: A zonegroup consists of multiple zones, this approximately
+  corresponds to what used to be called as a region in pre Jewel releases for
+  federated deployments. There should be a master zonegroup that will handle
```
Would this section just replace the section about federated deployments at some point in time? Otherwise that should at least get a warning that it doesn't work anymore with Jewel, it fails when there is no realm defined.

jrosenboom commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,710 @@
+
+::
+
+   # radosgw-admin zone create --rgw-zonegroup=us --rgw-zone=us-east-1 --endpoints=http://rgw1:80 --access-key=$ZONE_ACCESS_KEY --secret=$ZONE_SECRET_KEY --default --master
```
You didn't define `$ZONE_ACCESS_KEY` earlier, should this be `$SYSTEM_ACCESS_KEY` instead?

theanalyst replied:
Ah thanks, yeah I'll fix this

cbodley commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,710 @@
+===========
+
+While configuring zones, RadosGW, expects creation of zone a system user, with
+an S3 like access and secret keys. This allows another radosgw instance to pull
```
> While configuring zones, RadosGW, expects creation of zone a system user, with an S3 like access and secret keys.

While configuring zones, RadosGW expects creation of a system user, with S3-like access and secret keys.

theanalyst replied:
Thanks, I'll get this

cbodley commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,710 @@
+zonegroup will be in control of the zonegroup map and propagate changes to the
+rest of the system. We will also set this zonegroup as the default, which allows
+explicitly mentioning the ``rgw-zonegroup`` switch for later commands. We also
+make this the default zonegroup
```
> We will also set this zonegroup as the default, which allows explicitly mentioning the ``rgw-zonegroup`` switch for later commands. We also make this the default zonegroup

We will also set this zonegroup as the default, which allows later commands to use this zonegroup without explicitly mentioning it with the ``rgw-zonegroup`` switch.

cbodley commented on doc/radosgw/multisite.rst:
```diff
@@ -0,0 +1,710 @@
+    "epoch": 2
+  }
+
+  # radosgw-admin period pull --url=http://rgw1:80 --acess-key=$SYSTEM_ACCESS_KEY --secret=$SYSTEM_SECRET_KEY
```
> --acess-key=

--access-key=

cbodley commented:
nits aside, the content looks great. i like how you added a third zone to show the difference between configuration on the same cluster vs. a separate cluster

theanalyst commented:
@cbodley thanks, addressed the comments, squashed the commits, learnt a lot by creating the doc as well :)
